Conjur/conventions/ruby-style-guide.md

```diff
@@ -24,10 +23,194 @@ This guide is intended to be a "living document" in the sense that it's expected
+    end
+    ```
+
+  - Good:
+
+    ```ruby
+    def get_jwt_identity_from_request
+        @logger.debug(LogMessages::Authentication::AuthnJwt::CallingGetJwtIdentity.new)
+        jwt_identity
+        @logger.info(LogMessages::Authentication::AuthnJwt::FoundJwtIdentity.new(jwt_identity))
+    end
+    
+    def jwt_identity
+        @jwt_identity ||= @jwt_configuration.jwt_identity
+    end
+    ```
+
+- In constuctor of classes separate input handling and dependency handling with empty line
+
+- In command class implement methods in the order they are called
+
+  ```ruby
+  def call
+    func_a
+    func_b
+    func_c
+  end
+  
+  private
+  
+  def func_a
+    ...
+  end
+  
+  def func_b
+    ...
+  end
+  
+  def sub_func_b
+    ...
+  end
+  
+  def func_c
+    ...
+    end
+  ```
+
+- In functions with many parameters put each parameter in seperate line in the declaration. In this case end each param with ":" so the caller will need to specify name of variable for each param.
+
+  ```
+  def method(
+    param1:
+    param2:
+    param3:
+  )
+  ```
+
+  
+
+- When overriding a code climate violation with an inline exception (this isn't possible for all code climate  plugins but is for many), add the reason for the exception on the line above the special "ignore" directive
+
+   https://github.com/cyberark/conjur/blob/5fe6b22f225a1f2cf0be90d90f184c74c150be5a/cucumber/policy/features/support/rest_client_wrapper.rb#L12
+
+## Design Baseline
+
+- On any new class ask yourself if it should be a class or command class. Command Classes coordinate the execution of a multi-step process.
+
+  https://github.com/cyberark/conjur/blob/master/app/domain/README.md#the-functional-object-oriented-style
+
+  
+
+- No need for interface classes
+
+- *Never* use inheritance or mixins.
+
+- Dependencies for classes should be passed as default values in constructor
+
+  ```ruby
+  def initialize(
+    authenticator_input:,
+    logger: Rails.logger,
+    validate_resource_restrictions_class: Authentication::ResourceRestrictions::ValidateResourceRestrictions,
+    extract_token_from_credentials: Authentication::AuthnJwt::InputValidation::ExtractTokenFromCredentials.new,
+    create_identity_provider: Authentication::AuthnJwt::IdentityProviders::CreateIdentityProvider.new
+  )
+  ```
+
+- Use memoization when you need to run "hard operation" more than one and you sure that the output is the same (DB queries, API requests, etc..) . Use the '||=' to do it. The name of the function in memoization and the name of the memoaized object should be the same.
+
+  ```ruby
+  def identity_role
+    @identity_role ||= @role_class.by_login(
+      jwt_identity,
+      account: account
+    )
+  end
+  ```
+
+- When you want to run code that assignes value to class member in the first time it runs and afterwards it just returns it you need to check if you already run it, and if so to return the class member.
```

Comment:
I don't agree this guidance. I've seen this create a lot of confusion in the flow of code. 

In OOP, it's entirely appropriate for state to be maintained inside an object. In a functional programming, all methods really should be stateless (95% of the time - the exception is memoization).  Instead, just use memoization:

```ruby
def user_name
  expensive_api_call[:name]
end

# returns a hash
def expensive_api_call(client:, query:)
  @expensive_api_call ||= client.fetch(query)
end
```

I also want to clarify what the `@var ||= action` actually does. It performs a truthy test the value of the variable (`@var`).  If `@var` is set, it is returned.  If `@var` is not set, `action` is assigned to `@var` and returned (as the result of the last line of a method is returned by default).

Reply:
Not sure if i understand. If i have hard boolean operation that i want to memoaize call to it. And the return value is false. In case i don't check the memoaized value is false it will run again .

```
def boolean_hard_operation?
    # Perform api calls
    # Perform db queries
end
```

Reply:
@tzheleznyak, in that case, I'd recommend caching the responses of the above API and DB calls:

```ruby 
def boolean_hard_operation?
  # Perform api calls
  api_call 
  # Perform db queries
  db_call
end

def api_call
  @api_call ||= make_expensive_api_call
end

def db_call
  @db_call ||= make_expensive_db_call
end
```

The boolean logic is still calculated, but we're not focused on memoizing the boolean value (it's not the boolean calculation that is slow, it's the api and db calls required to by the boolean calculation that is slow).